lib/pavilion/system_plugins.py

```diff
@@ -0,0 +1,106 @@
+import collections
+from pavilion.variables import VariableSetManager
+from yapsy.IPlugin import IPlugin
+import logging
+import re
+
+LOGGER = logging.getLogger('pav.{}'.format(__name__))
+
+class PluginSystemError(RuntimeError):
+    pass
+
+_SYSTEM_PLUGINS = {}
+
+class SysVarDict( collections.UserDict ):
+
+    def __init__( self, want_deferred=True ):
+        global._SYSTEM_PLUGINS
+        super().__init__(_SYSTEM_PLUGINS)
+        self.want_deferred = want_deferred
+
+    def __getitem__( self, name ):
+        plugin = self.data[ name ]
+        return plugin.get()
+
+def add_system_plugin( system_plugin ):
+    name = system_plugin.name
+
+    if name not in _SYSTEM_PLUGINS:
+        _SYSTEM_PLUGINS[ name ] = system_plugin
+    elif priority > _SYSTEM_PLUGINS[name].priority:
+        _SYSTEM_PLUGINS[ name ] = system_plugin
+    elif priority == _SYSTEM_PLUGINS[name].priority:
+        raise PluginSystemError("Two plugins for the same system plugin have "
+                                "the same priority {}, {}."
+                                .format(system_plugin, _SYSTEM_PLUGINS[name]))
+
+def remove_system_plugin( system_plugin ):
+    name = system_plugin.name
+
+    if name in _SYSTEM_PLUGINS:
+        del _SYSTEM_PLUGINS[ name ]
+
+def get_system_plugin( name ):
+    if name not in _SYSTEM_PLUGINS:
+        raise PluginSystemError("Module not found: '{}'".format(name))
+
+    return _SYSTEM_PLUGINS[ name ]
+
+class SystemPlugin(IPlugin):
+
+    PRIO_DEFAULT = 0
+    PRIO_COMMON = 10
+    PRIO_USER = 20
+    global._SYSTEM_PLUGINS
+
+    def __init__(self, plugin_name, priority=PRIO_DEFAULT, is_deferable=False,
+                 sub_keys=[ None ]):
```

Comment:
Ok, this is the sub_keys problem I alluded to above. You probably want an empty list here, not a list containing a single [None] item that you later have to interpret. Unfortunately, even that's wrong.

When you have an keyword argument that should be a list:
 - **NEVER** initialize it in the function definition. That list is defined at module load time. If you modify it, the changes will carry over to every other call of the function. It is a terrible problem to debug, even when you know it can happen.
 - So if you can't create the list in the definition, what do you do? Set the default to None, then check for None and generate the true default in the function. This applies even when you're dealing with an empty list.

Reply:
Those two points apply to any mutable type, not just lists.